File: oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

public enum MapDirection {
    NORTH,
    EAST,
    SOUTH,
    WEST;

    @Override
    public String toString() {
        return switch (this) {
            case NORTH -> "N";
            case EAST -> "E";
            case SOUTH -> "S";
            case WEST -> "W";
        };
    }

    public MapDirection next() {
        int nextOrdinal = (this.ordinal() + 1) % values().length;
        return values()[nextOrdinal];
    }

    public MapDirection previous() {
        int previousOrdinal = (this.ordinal() + values().length - 1) % values().length;
        return values()[previousOrdinal];
    }
    public Vector2d toUnitVector() {
        return switch (this) {
            case NORTH -> new Vector2d(0, 1);
            case SOUTH -> new Vector2d(0, -1);
            case WEST -> new Vector2d(-1, 0);
            case EAST -> new Vector2d(1, 0);
        };
    }
}

File: oolab/kotlin/src/main/kotlin/model/MapDirection.kt
package model

enum class MapDirection {
    NORTH,
    EAST,
    SOUTH,
    WEST;

    override fun toString(): String {
        return when (this) {
            NORTH -> "N"
            EAST -> "E"
            SOUTH -> "S"
            WEST -> "W"
        }
    }

    fun next(): MapDirection {
        val nextOrdinal = (this.ordinal + 1) % entries.size
        return entries[nextOrdinal]
    }

    fun previous(): MapDirection {
        val previousOrdinal = (this.ordinal + entries.size - 1) % entries.size
        return entries[previousOrdinal]
    }

    companion object {
        fun random(): MapDirection {
            val randomOrdinal = (0..<entries.size).random()
            return entries[randomOrdinal]
        }
    }

}
